rfc/0075-faster-failover-and-configuration-push.md

```diff
@@ -0,0 +1,372 @@
+configuration attached if necessary, regardless the order of the operations.
+
+Several modifications are required in the SDK:
+1. Response handler should tolerate empty response with `NotMyVbucket` status, as the KV engine assumes that the SDK
+   already seen configuration, and there is no newer configuration available. In this case SDK should just retry the
+   operation.
+
+2. If the reponse payload contains body, it contains current configuration, which should be sent to configuration
+   monitor (manager). The SDK should either synchronously apply configuration, create waiting queue for given
+   `epoch`/`revision` pair.
+   Once configuration applied, the SDK must check if new configuration routes the operation to new endpoint or new
+   vbucket on the old endpoint, and *immediately* dispatch operation to new endpoint (or same endpoint in case vbucketID
+   has changed). In any other case, the SDK should send operation to retry orchestrator.
+   ```mermaid
+   flowchart 
+       A(NotMyVbucket) --> B{Empty Body?}
+       B -->|No|C(Apply Configuration)
+       C --> D{Route Operation}
+       D -->|Endpoint Changed| E[Dispatch To<br>New Endpoint]
+       D -->|VBucketID Changed| F[Update VBucketID]
+       F --> G[Dispatch To<br>Same Endpoint]
+       D -->|Everything Else| H[Send To Retry<br>Orchestrator]
+       B -->|Yes|H[Send To Retry<br>Orchestrator]
+   ```
+
+
+Below is a diagram that illustrates an example of the SDK workflow
+
+```mermaid
+sequenceDiagram
+    autonumber
+
+    participant conn_1
+    participant kv_node_1
+    participant retry_orchestrator
+    participant config_monitor
+
+            conn_1    ->>+  kv_node_1:       get("foo", vb=115)
+            kv_node_1 ->>-  conn_1:          NotMyVbucket(config={epoch=1, rev=11, ...})
+
+            conn_1   -->>+  config_monitor: propose confg={epoch=1, rev=11}
+
+
+    critical Check if config route operation to different node of vbucket
+
+        option "foo" still maps to kv_node_1
+            conn_1  -->>+  retry_orchestrator:  retry(get, "foo", reason=NotMyVbucket, epoch=1, rev=11)
+
+        option "foo" does not map to kv_node_1, or vbucket has changed 
+            conn_1    -->+  kv_node_1:       get("foo", vb=new_vbucket)
+            kv_node_1 ->>-  conn_1:          Success()
+    end
+```
+
+# Language Specifics
+
+## Feature Checklist
+
+1. `GetClusterConfigWithKnownVersion` (`0x1d`). The SDK should always supply current configuration version if the
+   connection has acknowledged feature flag.
+
+2. `DedupeNotMyVbucketClustermap` (`0x1e`). The SDK should be ready that the KV engine will not repeat configuration
+   payload if it already been sent to the socket by any means (`NotMyVbucket` status, `ClustermapChangeNotification`,
+   `GetClusterConfig`).
+
+3. Out-of-Order Execution. `Duplex` (`0x0c`) feature should be always negotiated in HELLO.
```

Comment:
Would it be better to call this "Server Push" instead of "Out-of-Order Execution"? It's possible to enable Duplex mode without also enabling out-of-order execution.

Reply:
You right, here I mean bi-direction communication `Duplex` (`0x0c`), which is not the same as Out-of-Order execution `UnorderedExecution` (`0x0e`). I will update the patch.